core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataDeleteIndexService.java

```diff
@@ -93,13 +105,23 @@ public ClusterState execute(final ClusterState currentState) {
                 MetaData.Builder metaDataBuilder = MetaData.builder(meta);
                 ClusterBlocks.Builder clusterBlocksBuilder = ClusterBlocks.builder().blocks(currentState.blocks());
 
+                final IndexGraveyard.Builder graveyardBuilder = IndexGraveyard.builder(metaDataBuilder.indexGraveyard());
+                final int previousGraveyardSize = graveyardBuilder.tombstones().size();
                 for (final Index index : indices) {
                     String indexName = index.getName();
                     logger.debug("[{}] deleting index", index);
                     routingTableBuilder.remove(indexName);
                     clusterBlocksBuilder.removeIndexBlocks(indexName);
                     metaDataBuilder.remove(indexName);
+                    graveyardBuilder.addTombstone(index); // add the index tombstone to the cluster state
                 }
+                final int numPurged = graveyardBuilder.purge(SETTING_MAX_TOMBSTONES.get(settings)); // purge tombstone entries if needed
```

Comment:
@dakrone I'm not sure I follow.  The number of tombstones we add here does not correlate to the number we may need to remove from the list to prevent it from growing too large.  If we add 10 tombstones in a single request, but have not reached our maximum limit that we can keep, then nothing should be deleted from the tombstones list.  

We do not want to delete tombstones until we are really compelled to (i.e. its growing too large).  That's the purpose of `purge`.  The reason the maximum tombstones exists is because we want to make that configurable, so if we start out with allowing 500 tombstones to remain in the cluster state, but at some point in the life of the cluster, someone decides they don't want so many tombstones to stick around for so long, an easy solution is to just drop the value of max tombstones, and on the next index deletion, the cluster state tombstones will be pruned to meet the new maximum value.

Not sure if this clarifies?